cl_manycore/libraries/bsg_manycore_cuda.cpp

```diff
@@ -47,312 +56,858 @@ static int hb_mc_get_y(tile_t *tiles, uint8_t *y_list, uint32_t num_tiles) {
+ * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
+ */
+int hb_mc_grid_init (device_t *device, hb_mc_dimension_t grid_dim, hb_mc_dimension_t tg_dim, char* name, uint32_t argc, uint32_t argv[]) {
+	int error; 
+	for (hb_mc_idx_t tg_id_x = 0; tg_id_x < grid_dim.x; tg_id_x ++) { 
+		for (hb_mc_idx_t tg_id_y = 0; tg_id_y < grid_dim.y; tg_id_y ++) { 
+			hb_mc_coordinate_t tg_id = hb_mc_coordinate(tg_id_x, tg_id_y);
+			error = hb_mc_tile_group_enqueue(device, device->num_grids, tg_id, grid_dim, tg_dim, name, argc, argv); 
+			if (error != HB_MC_SUCCESS) { 
+				bsg_pr_err("%s: failed to initialize tile group (%d,%d) of grid %d.\n", __func__, tg_id_x, tg_id_y, device->num_grids);
+				return HB_MC_FAIL;
+			}
+		}
+	}
+	device->num_grids ++;
+	return HB_MC_SUCCESS;
 }
 
-/*
- * Initializes Manycore tiles so that they may run kernels.
- * @param fd userspace file descriptor, which must be obtained from hb_mc_init_host.
- * @param eva_id specifies what the EVA-NPA mapping is.
- * @param tiles an array of tile_t structs to initialize.
- * @param num_tiles the number of tiles to initialize.
+
+
+
+/**
+ * Searches for a free tile group inside the device mesh and allocoates it, and sets the dimensions, origin, and id of tile group.
+ * @param[in]  device        Pointer to device
+ * @param[in]  tg            Pointer to tile group
  * @return HB_MC_SUCCESS on success and HB_MC_FAIL on failure. 
  */
-int hb_mc_device_init (uint8_t *fd, eva_id_t eva_id, char *elf, tile_t *tiles, uint32_t num_tiles) {
-	
-	int error = hb_mc_fifo_init(fd); 
-	if (error != HB_MC_SUCCESS) {
-		fprintf(stderr, "hb_mc_device_init() --> hb_mc_init_host(): failed to initialize host.\n");
+int hb_mc_tile_group_allocate_tiles (device_t *device, tile_group_t *tg){
+	int error;
+	if (tg->dim.x > device->mesh->dim.x){
+		bsg_pr_err("%s: tile group X dimension (%d) larger than mesh X dimension (%d).\n", __func__, tg->dim.x, device->mesh->dim.x);
 		return HB_MC_FAIL;
 	}
-	
-	if (eva_id != 0) {
-		return HB_MC_FAIL; /* eva_id not supported */
-	} 
-	
-	for (int i = 0; i < num_tiles; i++) { /* initialize tiles */
-		hb_mc_tile_freeze_dep(*fd, tiles[i].x, tiles[i].y);
-		hb_mc_tile_set_group_origin(*fd, tiles[i].x, tiles[i].y, tiles[i].origin_x, tiles[i].origin_y);
+	if (tg->dim.y > device->mesh->dim.y){
+		bsg_pr_err("%s: tile group Y dimension (%d) larger than mesh Y dimension (%d).\n", __func__, tg->dim.y, device->mesh->dim.y);
+		return HB_MC_FAIL;
 	}
+	for (hb_mc_idx_t org_y = device->mesh->origin.y; org_y <= (device->mesh->origin.y + device->mesh->dim.y - tg->dim.y); org_y++){
+		for (hb_mc_idx_t org_x = device->mesh->origin.x; org_x <= (device->mesh->origin.x + device->mesh->dim.x - tg->dim.x); org_x++){
+			int free = 1;
+			hb_mc_idx_t tile_id;
+			for (hb_mc_idx_t x = org_x; x < org_x + tg->dim.x; x++){
+				for (hb_mc_idx_t y = org_y; y < org_y + tg->dim.y; y++){
+					tile_id = (y - device->mesh->origin.y) * device->mesh->dim.x + (x - device->mesh->origin.x);
```

Comment:
I recommend making lines of code like this a function. It looks like you do this exact thing in a few different places. You'll thank yourself later.

Reply:
Yeah that makes sense, I should thank you! I added the following, what do you think? 
hb_mc_get_relative_coordinate (origin, coord): takes an absolute coordinate and an origin and subtract to return the relative coord. 
hb_mc_coordinate_to_flat_index (coord, dim): converts 2D coord to 1D index
hb_mc_dimension_to_size (dim) : converts 2D dimension to 1D size.

Reply:
What do flat_index and _size mean? 

Max and I typically reserve size for things referring to bytes.

Reply:
I think what you know mean by "flat index" (it's a 2D array flattened to 1D) but you can just call it an index -- after all that's what it is.
@drichmond is right -- we like to use `size` to mean "size in bytes" like the `sizeof` operator in C. Maybe just use `length`?
So is the plan to implement `hb_mc_coordinate_to_index()` in terms of `hb_mc_dimension_to_length()` and `hb_mc_get_relative_coordinate()`?

Reply:
Wow I should hit refresh every once in a while, just saw your replies! 
@drichmond, so as Max said, the flat_index is the conversion of 2D coordinates to 1D index, and you're right I will just call it index. And I will also change size to length.
@max-ruttenberg  actually I implemented it the other way around, so 'hb_mc_coordinate_to_index' just flattens a coordinate, and 'hb_mc_dimension_to_length' had nothing to do with flattening, I just built it to replace anywhere in code where is use dim.x * dim.y to get the total number of tiles. 
I added another function as you suggested called 'hb_mc_get_tile_id (origin, dim, coord) that uses 'hb_mc_get_relative_coordinate' and 'hb_mc_coordinate_index' to get the final tile id. Does that sound good?

Reply:
yeah that sounds good.